tpm/structures.go

```diff
@@ -231,73 +230,29 @@ type key12 struct {
 	KeyFlags        uint32
 	AuthDataUsage   byte
 	AlgorithmParams keyParams
-	PCRInfo         []byte // This must be a serialization of a pcrInfoLong.
-	PubKey          []byte
-	EncData         []byte
+	PCRInfo         tpmutil.U32Bytes // This must be a serialization of a pcrInfoLong.
+	PubKey          tpmutil.U32Bytes
+	EncData         tpmutil.U32Bytes
 }
 
 // A pubKey represents a public key known to the TPM.
 type pubKey struct {
 	AlgorithmParams keyParams
-	Key             []byte
+	Key             tpmutil.U32Bytes
 }
 
 // A symKey is a TPM representation of a symmetric key.
 type symKey struct {
 	AlgID     uint32
 	EncScheme uint16
-	Key       []byte
-}
-
-// TPMMarshal will marshal the symKey into a byte Buffer as specified the TPM spec.
-func (k *symKey) TPMMarshal(out io.Writer) error {
-	if err := binary.Write(out, binary.BigEndian, &k.AlgID); err != nil {
-		return err
-	}
-	if err := binary.Write(out, binary.BigEndian, &k.EncScheme); err != nil {
-		return err
-	}
-	if err := binary.Write(out, binary.BigEndian, uint16(len(k.Key))); err != nil {
-		return err
-	}
-	if err := binary.Write(out, binary.BigEndian, k.Key); err != nil {
-		return err
-	}
-	return nil
-}
-
-// TPMUnmarshal will parse the byte Buffer and populate the symKey.
-func (k *symKey) TPMUnmarshal(in io.Reader) error {
-	if err := binary.Read(in, binary.BigEndian, &k.AlgID); err != nil {
-		return err
-	}
-	if err := binary.Read(in, binary.BigEndian, &k.EncScheme); err != nil {
-		return err
-	}
-	var size uint16
-	if err := binary.Read(in, binary.BigEndian, &size); err != nil {
-		return err
-	}
-	key := make([]byte, size)
-	if err := binary.Read(in, binary.BigEndian, &key); err != nil {
-		return err
-	}
-	k.Key = key
-	return nil
-}
-
-// TPMPackedSize will return the expected size in bytes of the marshaled symKey.
-func (k *symKey) TPMPackedSize() int {
-	// AlgID  + EncScheme + len(Key) + Key
-	// uint32 + uint16    + uint16   + BYTE[]
-	return binary.Size(k.AlgID) + binary.Size(k.EncScheme) + binary.Size(uint16(len(k.Key))) + binary.Size(k.Key)
+	Key       tpmutil.U16Bytes
```

Comment:
Add a comment to clarify why this field is special

Reply:
Not sure what you mean by special, the TPM1.2 spec happens to use a 16-bit length for that byte array (I don't think there's any particular reason why).

Reply:
That's exactly what the comment should say.
Otherwise the next person reading this code will think "hmm, all the other byte slices here use U32Bytes, this one is U16Bytes. must be a bug, let me fix that..."